### PR TITLE
Parsing of  data from the M3 Sonar API to ROS2 PointCloud2 for usage in autonomous systems.

### LINKED ISSUE #1: [TASK] M3 Sonar Driver
## Description
Create a Sonar driver for ROS2

## Suggested Workflow
- [x] Set up a meeting with Vegard Haraldstad. He has volunteered to be a guide when setting up the sonar.
- [x] Get the Linux driver from either @KristianSem or Kongsberg. Try the one in Teams (Software/Sonar/M3 Sonar)
- [x] Run the Sonar using the Windows M3 app
- [x] Read API documentation and Setup documentation for Odroid XU4
- [x] Try connecting to the sonar with TCP (e.g. with PuTTY) on ubuntu
- [x] Create a new repo  
- [x] Set up a TCP connection in C++
- [x] Create a parser for the [IMB data format](https://mesotech.online/manuals/92220007002_m3_imb_beamformed_data_format_us_en.pdf). Asking someone from @vortexntnu/embedded-acoustics could be a good idea
- [ ] Make the driver

## Contacts
- Vegard Haraldstad - Mentor, He worked with sonars during his masters. Knows the data processing part well.
- @KristianSem - Mentor, He has experience testing the sonar and is a networking guru.

### PR DESCRIPTION
**Pull request by iTroned**
Closes #1 

Created a client connecting to the API with at a socket, reading the data as it arrives. Another thread requests this data at a chosen interval, parses the data to a point cloud and publishes it on a chosen topic.

### CODE REVIEW

File: src/imb/M3Publisher.cpp [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        }
+
+        data_vector_ = std::vector<uint8_t>();
+        packet_ready_ = false;
+        std::thread(&M3Publisher::CreateListener, this, this->get_parameter("ip_addr").as_string(), this->get_parameter("port").as_int()).detach();
+        timer_ = this->create_wall_timer(std::chrono::milliseconds(this->get_parameter("callback_time").as_int()), std::bind(&M3Publisher::ProcessData, this)); // How often to sample data from listener
+    }
+    void M3Publisher::ProcessData() {
+        try{
+            if(packet_ready_){
```
Comment (EirikKolas) [inline src/imb/M3Publisher.cpp:30]: replace with `if (!packet_ready_) return;` to avoid nesting

File: src/imb/imbFormat.cpp
Comment (EirikKolas) [inline src/imb/imbFormat.cpp:1]: remove commented out code

File: src/imb/M3Publisher.cpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                        float intensity = abs(data_body.complexData(i, j));
+                        if(intensity > 0){
+                            point.intensity = intensity;
+                        }
+                        else{
+                            point.intensity = 0;
+                        }
+                        cloud.points[i * data_header.nNumImageSample + j] = point;
+                    }
+                }
```
Comment (EirikKolas) [inline src/imb/M3Publisher.cpp:66]: Not a fan of the math being calculated in the ros node class. Could this be moved to the `imb::DataBody` class perhaps?

File: src/utils/RingBuffer.hpp
Comment (EirikKolas) [inline src/utils/RingBuffer.hpp:1]: Delete the ringbuffer if it isn't used

File: test/ImbFormatTest.cpp [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+    // ASSERT_EQ(dh.b3DScanAxis              , 0U);
+    // ASSERT_EQ(dh.bSonarClassID            , 0U);
+    // ASSERT_EQ(dh.byTX_Enabled             , 0U);
+    // ASSERT_EQ(dh.bReserved                , 0U);
+    // ASSERT_EQ(dh.fROV_Altitude            , 0U);
+    // ASSERT_EQ(dh.fConductivity            , 0U);
+    // ASSERT_EQ(dh.Reserved                 , 0U);
+    // clang-format on
+    
+}
```
Comment (EirikKolas) [inline test/ImbFormatTest.cpp:338]: Either finish the test or remove it :)

**Review by EirikKolas — CHANGES_REQUESTED** [review]
Great work. Only a couple minor code quality things I'm pedantic about

File: src/utils/netUtils.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    std::string processed_hex_string = hex_string;
+    processed_hex_string.erase(std::remove_if(processed_hex_string.begin(), processed_hex_string.end(), [](unsigned char c){
+        return std::isspace(c);
+    }), processed_hex_string.end());
+
+    // Ensure the processed string has an even length
+    if (processed_hex_string.length() % 2 != 0) {
+        // handle error, e.g., throw exception, return, assert, etc.
+        // throw std::invalid_argument("Invalid hex string length after removing whitespace. Must be even.");
+        // return;
```
Comment (alekskl01) [inline src/utils/netUtils.cpp:39]: comment back in or remove

File: test/M3SonarTest.cpp
Comment (alekskl01) [inline test/M3SonarTest.cpp:1]: remove commented out code

**Review by EirikKolas — APPROVED** [review]

**Pull request merged**